esp32c3 interrupt implementation

zdima opened the pull request:
@aykevl this is only interrupt implementation for esp32c3.
No exported APIs except enum values of `HandlerID` type to initialize chip's modules.
Tested with pin src/examples/pininterrupt/pininterrupt.go (not checked in)

zdima committed:
esp32c3 interrupts

zdima committed:
cleanup

zdima committed:
move to runtime to avoid export of Init

aykevl commented on src/runtime/interrupt/interrupt_esp32c3.go:
```diff
@@ -0,0 +1,198 @@
+	}
+}
+
+//export handleException
```
Why is this exported?

zdima replied:
will remove it was part of the interrupt table

aykevl commented on src/runtime/runtime_esp32c3.go:
```diff
@@ -63,3 +68,28 @@ func abort() {
 	}
 }
+
+// Init initialize the interrupt controller and called from runtime once.
```
```suggestion
// interruptInit initialize the interrupt controller and called from runtime once.
```

aykevl commented on src/runtime/interrupt/interrupt_esp32c3.go:
```diff
@@ -0,0 +1,198 @@
+		esp.INTERRUPT_CORE0.CPU_INT_ENABLE.ClearBits(interruptBit)
+
+		// reset this interrupt
+		if esp.INTERRUPT_CORE0.CPU_INT_TYPE.Get()&(interruptBit) != 0 {
```
The extra parentheses are unnecessary.

```suggestion
		if esp.INTERRUPT_CORE0.CPU_INT_TYPE.Get() & interruptBit != 0 {
```

zdima commented on src/runtime/interrupt/interrupt_esp32c3.go:
```diff
@@ -0,0 +1,198 @@
+func getHandlerIDsForInterrupt(interrupt int) []HandlerID {
+	switch interrupt {
+	case 1:
+		return []HandlerID{GPIO}
```
@aykevl, is this bad in terms of allocation?

niaow replied:
You could probably fix this by moving the arrays into globals.

zdima replied:
Thanks, move to global.

zdima committed:
mepc must be preserved and restored before mret.
the mpie must be set before mret.
remove memory allocation from interrupt code.

aykevl commented on src/runtime/interrupt/interrupt_esp32c3.go:
```diff
@@ -0,0 +1,173 @@
+//export handleException
+func handleException(mcause uintptr) {
+	// TODO need to get location of actual MEPC from the stack stash created in src/device/riscv/handleinterrupt.S
+	println("*** Exception:     pc:", riscv.MEPC.Get())
```
I don't understand the TODO, isn't this the actual MEPC?

zdima replied:
I don't know how to access the saved registers on the stack in handleinterrupt.S.
I want to print all the registers here and I can add it If you can suggest how I can access location of the stack created by handleinterrupt.S

zdima replied:
removed

aykevl reviewed (commented):
I've been thinking about this PR and how to implement interrupts correctly on the ESP32-C3.
I think it would actually make more sense to make the `Interrupt` type be the 31 interrupt vectors. It would then work as follows:

* Code that wants to use an interrupt can do so using `interrupt.New`, for example `interrupt.New(5, func(interrupt.Interrupt) { ... })`.
* It will have to manually map the interrupt source to the interrupt vector, for example `esp.INTERRUPT_CORE0.GPIO_INTERRUPT_PRO_MAP.Set(5)`.
* The `handleInterrupt` logic now becomes a lot simpler. It can now directly call `callInterruptHandler` with the interrupt number from `mcause`.

It is still possible to map multiple interrupt sources to a single interrupt vector: `interrupt.New` can be used for multiple handlers on a single interrupt number. For example:

```go
// configuring GPIO:
intr := interrupt.New(5, func(interrupt.Interrupt) {
    status := esp.GPIO.STATUS.Get()
    if status != 0 {
        // handle GPIO interrupt
    }
})
intr.Enable()
esp.INTERRUPT_CORE0.GPIO_INTERRUPT_PRO_MAP.Set(5)

// configuring UART:
intr := interrupt.New(5, func(i interrupt.Interrupt) {
    flag := esp.UART0.INT_ST.Get()
    if flag != 0 {
        // handle UART interrupt
    }
})
intr.Enable()
esp.INTERRUPT_CORE0.UART0_INTERRUPT_PRO_MAP.Set(5)
```

(These two don't need to be on the same interrupt vector, I'm only showing that it is possible).

What do you think?

zdima commented:
I didn't know you can call `interrupt.New` with the same id. I never get response from you on how does it work in the slack PM :(. If I knew, perhaps the implementation will be much simpler.
So how does it work if make a call `interrupt.New(5, func(interrupt.Interrupt) {})` twice?

zdima commented:
@aykevl not sure if you get notification on my comments above

zdima committed:
fix function comments

zdima committed:
remove export

zdima committed:
remove extra parentheses

zdima committed:
add support for gdb on Apple M1 running in i386 mode

zdima committed:
assume the interrupt.Interrupt value is equal to CPU interrupt number.

zdima committed:
remove TODO since we not printing registers

aykevl reviewed (approved):
Looks good to me, thank you for all the work!

aykevl commented:
CI is failing, but I think that's an unrelated issue.

aykevl commented:
Oh now I see. You have made the PR against the espnet branch. Can you rebase it against the dev branch?

zdima commented:
I could, but don’t we need other changes you made like stdio, pico lib and FreeRTOS?
 

> On Oct 22, 2021, at 6:59 PM, Ayke ***@***.***> wrote:
> 
> 
> Oh now I see. You have made the PR against the espnet branch. Can you rebase it against the dev branch?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/tinygo-org/tinygo/pull/2170#issuecomment-950003303>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACI2FGP4NA36ES4CPIN3L3UIHUDVANCNFSM5F2XORKA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>

aykevl commented:
> I could, but don’t we need other changes you made like stdio, pico lib and FreeRTOS?

Yes, but by merging this in the dev branch it will become available to everybody, not just those working from the espnet branch. That's why I also made a few other PRs like #2135, #2136, and #2143. In general I prefer working as much as possible directly from the dev branch for independent features.
After the PR is merged, it can easily be added to the espnet branch too.

zdima commented:
this is merged with PR #2200
